go/p2p/rpc/client.go

```diff
@@ -374,37 +415,38 @@ func (c *client) CallMulti(
 			}
 
 			rsp := reflect.New(reflect.TypeOf(rspTyp)).Interface()
-			pf, err := c.call(peerCtx, peer, &request, rsp, maxPeerResponseTime)
-			ch.In() <- &result{rsp, pf, err}
-		})
-	}
+			pf, err := c.timeCall(peerCtx, peer, &request, rsp, co.maxPeerResponseTime)
 
-	if len(resultChs) == 0 {
-		return nil, nil, nil
+			resultCh <- result{rsp, pf, err}
+		})
 	}
-	resultCh := channels.NewNativeChannel(channels.None)
-	channels.Multiplex(resultCh, resultChs...)
```

Comment:
Hm, the below construct is not equivalent as it waits in order so it will always wait for the first peer to complete etc. The `channels.Multiplex` basically does one large `select` over all of the channels, so it will process results as they come in, regardless of whether the last peer in the list actually produced results first. This is important  when the aggregation function asks to break early and not wait for all of the peers.

Reply:
Not sure if this is true as the channel is buffered/non blocking. Tested and I get a random order `[1 7 3 13 5]`.

```
package main

import (
	"context"
	"fmt"
	"math/rand"
	"time"

	"github.com/oasisprotocol/oasis-core/go/common/workerpool"
)

func main() {
	peers := []int{}
	for i := 0; i < 100; i++ {
		peers = append(peers, i)
	}

	pool := workerpool.New("p2p/rpc")
	pool.Resize(10)
	defer pool.Stop()

	peerCtx, cancel := context.WithCancel(context.Background())
	defer cancel()

	type result struct {
		rsp int
		err error
	}

	resultCh := make(chan result, len(peers))

	for _, peer := range peers {
		peer := peer

		pool.Submit(func() {
			select {
			case <-peerCtx.Done():
				return
			default:
			}

			var err error
			if peer%2 == 0 {
				err = fmt.Errorf("even peers fail")
			}

			time.Sleep(time.Duration(rand.Intn(1000)) * time.Millisecond)

			resultCh <- result{peer, err}
		})
	}

	var rsps []int

loop:
	for i := 0; i < len(peers); i++ {
		select {
		case result := <-resultCh:
			if result.err != nil {
				break
			}

			rsps = append(rsps, result.rsp)

			if len(rsps) == 5 {
				break loop // agregate function
			}

		case <-peerCtx.Done():
			break loop
		}
	}

	fmt.Printf("%+v\n", rsps)
}
```

Reply:
Imagine the following scenario:
* 10 peers in total.
* First 5 peers take 30 seconds to complete the request.
* Last 5 peers take 1 second to complete the request.
* Aggregation function signals to break after getting response from 5 peers.

In this case `CallMulti` should not wait 30 seconds to return as it would have the 5 results after roughly 1 second.

Based on how `Multiplex` is implemented (a `select` over the result channels of all peers at once) it should do the right thing while the new implementation would wait for result from the first peer, then the second, etc.

Reply:
Ah sorry for some reason I thought the result channel was per peer, you're right this looks ok.